File: src/game/src/turtleSnake.cpp
#include "ros/ros.h"
#include <string>
#include <cstdlib>
#include <iostream>
#include <bits/stdc++.h>
#include <ctime>
#include <random>
#include <math.h>
#include <typeinfo>

#include "turtlesim/Pose.h"
#include "turtlesim/Spawn.h"
#include "turtlesim/TeleportAbsolute.h"
#include "turtlesim/SetPen.h"
#include "turtlesim/Kill.h"
#include "std_srvs/Empty.h"
#include "geometry_msgs/Twist.h"

class Turtle{
    private:
        ros::NodeHandle nh;
        ros::Subscriber sub;

    public:
        turtlesim::Pose pose;
        std::string name;
        ros::Publisher pub;
        bool activeSub;

        Turtle(const std::string turtleName) {
            name = turtleName;
            sub = nh.subscribe("/"+name+"/pose", 100, &Turtle::updatePose , this);
            pub = nh.advertise<geometry_msgs::Twist>("/"+name+"/cmd_vel", 10);
            activeSub = false;
        }

        void updatePose(const turtlesim::Pose::ConstPtr& data)
        {
            this->pose = *data;
            activeSub = true;
        }

        void setZeroPen()
        {
            turtlesim::SetPen srv;
            ros::ServiceClient client = nh.serviceClient<turtlesim::SetPen>("/"+name+"/set_pen");
            client.waitForExistence();

            srv.request.off = 1;

            if (!client.call(srv))
            {
                ROS_ERROR("Failed to call service: /%s/set_pen", name.c_str());
            }
        }

        void spawn()
        {
            turtlesim::Spawn srv;
            ros::ServiceClient client = nh.serviceClient<turtlesim::Spawn>("/spawn");
            client.waitForExistence();
            // Variables needed to create floats random numbers for the angle
            std::random_device rd;
            std::default_random_engine gen(rd());
            std::uniform_real_distribution<double> distribution(0.0, 2*M_PI);
            
            srv.request.x = rand()%12; //Where 12 is the sixe of the board
            srv.request.y = rand()%12;
            srv.request.theta = distribution(gen);
            srv.request.name = name;

            if (!client.call(srv))
            {
                ROS_ERROR("Failed to call service: '/spawn'");
            }
            pose.x = srv.request.x;
            pose.y = srv.request.y;
            pose.theta = srv.request.theta;
            pose.linear_velocity = 0.0;
            pose.angular_velocity = 0.0;
            
            setZeroPen();

            std::cout << "Spawn new turtle: " << name << std::endl;
        }

        void moveToEnd(turtlesim::Pose &frontTurtlePose)
        {
            float deltaX = 0.6;
            float deltaY = 0.6;

            turtlesim::TeleportAbsolute srv;
            ros::ServiceClient client = nh.serviceClient<turtlesim::TeleportAbsolute>("/"+name+"/teleport_absolute");
            client.waitForExistence();


            srv.request.x = frontTurtlePose.x - deltaX*std::cos(frontTurtlePose.theta);
            srv.request.y = frontTurtlePose.y - deltaY*std::sin(frontTurtlePose.theta);
            srv.request.theta = frontTurtlePose.theta;

            if (!client.call(srv))
            {
                ROS_ERROR("Failed to call service: /%s/teleport_absolute", name.c_str());
            }
        }

        void stopTurtle()
        {
            geometry_msgs::Twist zeroVelTwist;
            pub.publish(zeroVelTwist);
        }
};


void changeBackground(float r=69, float g=86, float b=255)
{
    ros::NodeHandle cb;
    std_srvs::Empty srv;
    
    cb.setParam("/turtlesim/background_r", r);
    cb.setParam("/turtlesim/background_g", g);
    cb.setParam("/turtlesim/background_b", b);

    ros::ServiceClient client = cb.serviceClient<std_srvs::Empty>("/clear");
    client.waitForExistence();

    if (!client.call(srv))
    {
        ROS_ERROR("Failed to call service: '/clear'");
    }
}

void resetGame(ros::NodeHandle &rg)
{
    ros::ServiceClient client = rg.serviceClient<std_srvs::Empty>("/reset");
    client.waitForExistence();
    std_srvs::Empty srv;
    if (!client.call(srv))
    {
        ROS_ERROR("Failed to call service: '/reset'");
        
    }

    changeBackground();
}

bool isInsideArea(turtlesim::Pose*& turtle1Pose, turtlesim::Pose*& turtle2Pose)
{
    float deltaX{0.5}, deltaY{0.5};
    bool xCheck{turtle1Pose->x <= turtle2Pose->x + deltaX && turtle1Pose->x >= turtle2Pose->x - deltaX};
    bool yCheck{turtle1Pose->y <= turtle2Pose->y + deltaY && turtle1Pose->y >= turtle2Pose->y - deltaY};

    if (xCheck && yCheck)
        return true;
    else
        return false;

}

void stopSnake(std::vector<std::unique_ptr<Turtle>> &turtles)
{
    for (int i = 0; i < turtles.size()-1; i++)
    {
        turtles.at(i)->stopTurtle();
    }
}

bool hitTheWall(turtlesim::Pose*& headPose)
{
    std::vector<float> xSize{0.0, 11.05}, ySize{0.0, 11.05};

    if (headPose->x == xSize.at(0) || headPose->x >= xSize.at(1) || headPose->y == ySize.at(0) || headPose->y >= ySize.at(1))
        return true;
    else
        return false;
}

void killSnake(std::vector<std::unique_ptr<Turtle>> &turtles)
{
    ros::NodeHandle nh;

    turtlesim::Kill srv;
    ros::ServiceClient client = nh.serviceClient<turtlesim::Kill>("/kill");
    client.waitForExistence();
    
    for (int i=turtles.size()-1; i>=0; i--)
    {   
        srv.request.name = turtles.at(i)->name;
        if (!client.call(srv))
            ROS_ERROR("Failed to call service: /kill on %s", srv.request.name.c_str());

        ros::Duration(0.05).sleep();
    }
}

void startGame(ros::NodeHandle &nh)
{
    int snakeLength = 0;
    int maxLengthSnake = 200;
    double deltaPos;
    double angGoal;
    double deltaAng;
    geometry_msgs::Twist newVel;
    double freq{70}; // I decided to use a frequency of 70 because the subscriber subscribe with an average of 62Hz
    ros::Rate rate(freq);


    /* Vector containing all the spawned turtles. I tried to create a simple vector made of Turtle object, but I faced a problem
       where once instantiate the second object, the subscriber of the previous was dying, loosing the updation of the actual pose. */  
    std::vector<std::unique_ptr<Turtle>> turtles;

    turtles.emplace_back(std::make_unique<Turtle>("turtle1"));
    turtles.back()->setZeroPen();
    snakeLength = 1;

    turtles.emplace_back(std::make_unique<Turtle>("turtle2"));
    turtles.back()->spawn();
    // Here we do not increase snakeLength because the above turtle has been spawned, but it still not part of the snake itself
    
    // For better comprehension, saving the head pose as new name
    turtlesim::Pose* headPose{&(turtles[0]->pose)};

    while (ros::ok())
    {
        // Waiting the subscriber to be active
        while (!turtles.at(0)->activeSub)
        {
            ros::spinOnce();
        }

        turtlesim::Pose* targetTurtle{&(turtles.back()->pose)};

        if (isInsideArea(headPose, targetTurtle))
        {
            turtles[snakeLength]->moveToEnd(turtles[snakeLength-1]->pose);
            turtles.emplace_back(std::make_unique<Turtle>("turtle"+std::to_string(snakeLength+2)));
            turtles.back()->spawn();
            snakeLength += 1;
        } 

        if (hitTheWall(headPose))
        {
            changeBackground(255, 0, 0);
            stopSnake(turtles);
            killSnake(turtles);

            ROS_INFO("You hit the WALL! You LOST :(");
            ros::shutdown();
        }

        if (snakeLength > 1)
        {
            for(int i=1; i<snakeLength; i++)
            {
                deltaPos = sqrt(pow(turtles.at(i-1)->pose.x - turtles.at(i)->pose.x,2) + pow(turtles.at(i-1)->pose.y - turtles.at(i)->pose.y,2) );
                angGoal = atan2(turtles.at(i-1)->pose.y - turtles.at(i)->pose.y, turtles.at(i-1)->pose.x - turtles.at(i)->pose.x);
                deltaAng = atan2(sin(angGoal - turtles.at(i)->pose.theta), cos(angGoal - turtles.at(i)->pose.theta));

                if (deltaPos < 0.7)
                {
                    newVel.linear.x = 0;
                    newVel.angular.z = 0;
                }
                else
                {
                    int kl = 3;
                    int ka = 15;
                    newVel.linear.x = kl*deltaPos;
                    newVel.angular.z = ka*deltaAng;
                }
                turtles.at(i)->pub.publish(newVel);
            }
        }

        ros::spinOnce();
        rate.sleep();
    }   
}


int main(int argc, char **argv)
{
    // Making rundom number always change
    srand ( time(NULL) );

    // Initializing the node
    ros::init(argc, argv, "turtleSnake_game");
    ros::NodeHandle nh;

    // Resetting the board as default settings
    resetGame(nh);
    startGame(nh);
}

File: src/game/src/turtleSnake.py
#!/usr/bin/env python3

import rospy
import random
import math
from turtlesim.msg import Pose
from turtlesim.srv import Spawn, TeleportAbsolute, SetPen, Kill
from std_srvs.srv import Empty
from geometry_msgs.msg import Twist

turtles = []        # List containing all the turtle present on board
snake_lenght = 1

class Turtle:
    def __init__(self, name):
        self.name = name
        self.pose = Pose()

        rospy.Subscriber(f"/{name}/pose", Pose, self.update_pose)
        self.pub = rospy.Publisher(f"/{name}/cmd_vel", Twist, queue_size=10)

    def update_pose(self, data):
        self.pose = data

    def set_zero_pen(self):
        rospy.wait_for_service(f"/{self.name}/set_pen")
        zero_pen = rospy.ServiceProxy(f"/{self.name}/set_pen", SetPen)
        zero_pen(0, 0, 0, 0, 1)

    def spawn(self):
        rospy.wait_for_service("/spawn")
        spawn_new_turtle = rospy.ServiceProxy("/spawn", Spawn)
        rand_x = random.uniform(0.0, 11)
        rand_y = random.uniform(0.0, 11)
        rand_theta = random.uniform(0.0, 2*math.pi)
        spawn_new_turtle(rand_x, rand_y, rand_theta, self.name)

        # We set the pose manually since the subscriber can require more time than the execution code
        self.pose.x = rand_x
        self.pose.y = rand_y
        self.pose.theta = rand_theta
        self.pose.linear_velocity = 0.0
        self.pose.angular_velocity = 0.0

        self.set_zero_pen()
        print(f"Spawn new turtle: {self.name}")

    def move_to_end(self, front_turtle_pose):
        delta_x = 0.6
        delta_y = 0.6

        rospy.wait_for_service(f"/{self.name}/teleport_absolute")
        move_turtle = rospy.ServiceProxy(f"/{self.name}/teleport_absolute", TeleportAbsolute)

        move_turtle(front_turtle_pose.x - delta_x*math.cos(front_turtle_pose.theta),
                    front_turtle_pose.y - delta_y*math.sin(front_turtle_pose.theta),
                    front_turtle_pose.theta)
    
    def stop_turtle(self):
        zero_vel_twist = Twist()
        self.pub.publish(zero_vel_twist)


def is_inside_area(turtle1_pose, turtle2_pose):
    delta_x = 0.5
    delta_y = 0.5
    check_x = turtle1_pose.x <= turtle2_pose.x + delta_x and turtle1_pose.x >= turtle2_pose.x - delta_x
    check_y = turtle1_pose.y <= turtle2_pose.y + delta_y and turtle1_pose.y >= turtle2_pose.y - delta_y
    
    if check_x and check_y:
        return True
    else:
        return False

def reset_game():
    rospy.wait_for_service('/reset')
    reset_game = rospy.ServiceProxy("/reset", Empty)
    reset_game()
    change_background()

def hit_the_wall():
    x_size = [0, 11.05]
    y_size = [0, 11.05]
    if turtles[0].pose.x == x_size[0] or turtles[0].pose.x >= x_size[1] or turtles[0].pose.y == y_size[0] or turtles[0].pose.y >= y_size[1]:
        return True
    else:
        return False

def change_background(r=69, g=86, b=255):
    rospy.wait_for_service('/clear')

    rospy.set_param('/turtlesim/background_r' , r)
    rospy.set_param('/turtlesim/background_g' , g)
    rospy.set_param('/turtlesim/background_b' , b)

    change_bkg_color = rospy.ServiceProxy("/clear", Empty)
    change_bkg_color()

def kill_snake():
    rospy.wait_for_service('/kill')
    kill = rospy.ServiceProxy("/kill", Kill)

    for i in range(len(turtles)-1, -1, -1):
        kill(turtles[i].name)
        print(f"Killed turtle: {turtles[i].name}")
        rospy.sleep(0.05)
    snake_lenght == 0

def stop_snake():
    for i in range(len(turtles)-1):
        turtles[i].stop_turtle()


# Where the actual game is located
def callback(data):
    global turtles
    global snake_lenght

    #Updating head (=turtle1) pose
    turtles[0].pose = data
    
    # Reassignment only for a clearer variable name
    head_pose = turtles[0].pose
    new_vel = Twist()

    if is_inside_area(head_pose, turtles[-1].pose):
        # Teleport the target turtle to the end of the snake when we are "close enough"
        turtles[-1].move_to_end(turtles[-2].pose)

        y = Turtle(f"turtle{snake_lenght+2}")
        y.spawn()
        turtles.append(y)

        snake_lenght += 1
        
    if hit_the_wall():
        change_background(255, 0, 0)
        stop_snake()
        kill_snake()

        rospy.loginfo("You hit the WALL! You LOST :(")
        rospy.signal_shutdown("You Lost")

    if snake_lenght > 1:  
        for i in range(1, len(turtles)-1):
            delta_pos = math.sqrt(pow(turtles[i-1].pose.x - turtles[i].pose.x ,2) + pow(turtles[i-1].pose.y - turtles[i].pose.y,2))
            ang_goal = math.atan2(turtles[i-1].pose.y - turtles[i].pose.y, turtles[i-1].pose.x - turtles[i].pose.x)
            delta_ang = math.atan2(math.sin(ang_goal - turtles[i].pose.theta), math.cos(ang_goal - turtles[i].pose.theta))

            if delta_pos < 0.7:
                new_vel.linear.x = 0
                new_vel.angular.z = 0
            else:
                kl = 3                              # linear vel proportional coefficient
                ka = 15                             # angular vel proportional coefficient
                new_vel.linear.x = kl * delta_pos
                new_vel.angular.z =  ka * delta_ang

            turtles[i].pub.publish(new_vel)


if __name__ == "__main__":
    # Initializing the node
    rospy.init_node("turtleSnake_game", anonymous = False)

    # Resetting the board as default settings
    reset_game()

    # Creating the object for the first turtle (already spawned when runned the node: "turtlesim_node")
    turtle1 = Turtle("turtle1")
    turtle1.set_zero_pen()
    turtles.append(turtle1)
    
    # Spawning a new turtle
    turtle2 = Turtle("turtle2")
    turtle2.spawn()
    turtles.append(turtle2)

    rospy.Subscriber("/turtle1/pose", Pose, callback)

    rospy.spin()
